src/shims/fs.rs

```diff
@@ -257,6 +264,197 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                        return Ok(-1);
+                    }
+                }
+            } else {
+                if let Some(path) = &this
+                    .machine
+                    .file_handler
+                    .handles
+                    .get(&dirfd)
+                    .map(|h| &h.path)
+                {
+                    // If `dirfd` points to a directory, `pathname` is relative to that directory
+                    if path.is_dir() {
+                        path.join(pathname)
+                    } else {
+                        // If `pathname` is empty and the `AT_EMPTY_PATH` flag is set, we use the path of `dirfd`.
+                        if pathname.as_os_str().is_empty() && (flags & at_empty_path != 0) {
+                            (*path).clone()
+                        } else {
+                            // It is not specified what should happen here.
+                            throw_unsup_format!(
+                                "the file descriptor `{}` and path `{:?}` combination is invalid.",
+                                dirfd,
+                                path
+                            );
+                        }
+                    }
+                } else {
+                    return this.handle_not_found();
+                }
+            }
+        };
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
+            std::fs::symlink_metadata(path)
+        } else {
+            std::fs::metadata(path)
+        };
+
+        let metadata = match metadata {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        // The `mode` field specifies the type of the file and the permissions over the file for
+        // the owner, its group and other users. Given that we can only provide the file type
+        // without using platform specific methods, we only set the bits corresponding to the file
+        // type.
+        let mode = if file_type.is_file() {
+            this.eval_libc("S_IFREG")?
+        } else if file_type.is_dir() {
+            this.eval_libc("S_IFDIR")?
+        } else {
+            this.eval_libc("S_IFLNK")?
+        }
+        .to_u32()? as u16;
```

Comment:
Another lossy cast.

Reply:
I'm torn about this one. At first I did it because there was no `Scalar::to_u16` method.

According to `man statx`, the `stx_mode` should be an `__u16`. But in [libc](https://docs.rs/libc/0.2.63/libc/constant.S_IFLNK.html) all the constants are `mode_t`s (which is the same as `u32`).

Reply:
All the more reason to use `try_into().unwrap()`, so we at least notice if anyone starts breaking these expectations

Reply:
Or you add `Scalar::to_u16` to rustc :)

Reply:
Ok then :P